File: py/engit/_web.py
"""engit web command — open the current repository on GitHub.

Resolves the remote URL and opens it in the default web browser.
"""

from __future__ import annotations

import webbrowser
from pathlib import Path

from ._exceptions import GitError
from ._git import getCurrentBranch, getRemoteUrl, requireGitRepo


def _toHttpsUrl(remote_url: str) -> str:
    """Normalise an SSH or HTTPS remote URL to an HTTPS browser URL.

    Handles:
    * ``git@github.com:owner/repo.git`` → ``https://github.com/owner/repo``
    * ``https://github.com/owner/repo.git`` → ``https://github.com/owner/repo``
    * ``https://github.com/owner/repo`` → unchanged

    """
    url = remote_url.strip()

    # SSH format: git@host:owner/repo.git
    if url.startswith('git@') and ':' in url:
        host, path = url[len('git@') :].split(':', 1)
        url = f'https://{host}/{path}'

    # Strip trailing .git
    if url.endswith('.git'):
        url = url[:-4]

    return url


def runWeb(
    *,
    branch: str | None = None,
    remote: str = 'origin',
    cwd: Path | None = None,
) -> None:
    """Open the current repository on GitHub in a web browser.

    Args:
        branch: Branch or tag to view. Defaults to the currently checked-out
            branch. Pass ``None`` to open the repository root.
        remote: Remote whose URL is used. Defaults to ``'origin'``.
        cwd: Git working directory. Defaults to the current directory.

    Raises:
        ~._exceptions.NotAGitRepoError: If not inside a git repo.
        ~._exceptions.GitError: If the remote URL cannot be determined.

    """
    requireGitRepo(cwd=cwd)

    raw_url = getRemoteUrl(remote=remote, cwd=cwd)
    if not raw_url:
        raise GitError(f"Remote '{remote}' has no URL. Check your git remote configuration.")

    base_url = _toHttpsUrl(raw_url)

    # Append the branch/tree path if we have one
    resolved_branch = branch or getCurrentBranch(cwd=cwd)
    if resolved_branch:
        url = f'{base_url}/tree/{resolved_branch}'
    else:
        url = base_url

    print(f'Opening {url}')
    try:
        webbrowser.open(url, new=2)
    except webbrowser.Error as exc:
        raise GitError(f'Could not open web browser: {exc}') from exc

File: rust/engit-core/src/web.rs
//! Open the current repository on GitHub.

use std::path::Path;
use std::process::Command;

use crate::error::{EngitError, Result};
use crate::git::{get_current_branch, get_remote_url, require_git_repo};

pub(crate) fn to_https_url(remote_url: &str) -> String {
    let mut url = remote_url.trim().to_string();

    if let Some(rest) = url.strip_prefix("git@") {
        if let Some((host, path)) = rest.split_once(':') {
            url = format!("https://{host}/{path}");
        }
    }

    if url.ends_with(".git") {
        url.truncate(url.len() - 4);
    }

    url
}

fn open_url(url: &str) -> Result<()> {
    let status = if cfg!(windows) {
        Command::new("cmd").args(["/C", "start", "", url]).status()
    } else if cfg!(target_os = "macos") {
        Command::new("open").arg(url).status()
    } else {
        Command::new("xdg-open").arg(url).status()
    }
    .map_err(|source| EngitError::Git(format!("Could not open web browser: {source}")))?;

    if status.success() {
        Ok(())
    } else {
        Err(EngitError::Git(format!(
            "Could not open web browser: command exited with status {status}."
        )))
    }
}

/// Open the current repository on GitHub in a browser.
pub fn run_web(branch: Option<&str>, remote: &str, cwd: Option<&Path>) -> Result<()> {
    require_git_repo(cwd)?;

    let raw_url = get_remote_url(remote, cwd).ok_or_else(|| {
        EngitError::Git(format!(
            "Remote '{remote}' has no URL. Check your git remote configuration."
        ))
    })?;
    let base_url = to_https_url(&raw_url);
    let resolved_branch = branch
        .map(ToOwned::to_owned)
        .or_else(|| get_current_branch(cwd));
    let url = resolved_branch
        .map(|branch| format!("{base_url}/tree/{branch}"))
        .unwrap_or(base_url);

    println!("Opening {url}");
    open_url(&url)
}

#[cfg(test)]
mod tests {
    use super::to_https_url;

    #[test]
    fn normalizes_ssh_and_https_urls() {
        assert_eq!(
            to_https_url("git@github.com:owner/repo.git"),
            "https://github.com/owner/repo"
        );
        assert_eq!(
            to_https_url("https://github.com/owner/repo.git"),
            "https://github.com/owner/repo"
        );
    }
}
